src/v/storage/batch_consumer_utils.h

```diff
@@ -0,0 +1,162 @@
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "storage/parser.h"
+
+#include <seastar/core/future.hh>
+
+#include <exception>
+
+namespace storage {
+
+/// Chain multiple batch_consumer instances to consume
+/// from the same input_stream. The consumers are allowed to skip
+/// batches or stop early.
+///
+/// Every consumer in the chain observes the batches as if it's the
+/// only one in the chain. If one of the consumers skips batches or
+/// stops early it shouldn't affect the others.
+template<int N>
+class chained_batch_consumer : public storage::batch_consumer {
+public:
+    template<typename... Args>
+    explicit chained_batch_consumer(Args&&... args)
+      : _cons{std::forward<Args>(args)...} {
+        std::ignore = std::fill_n(
+          _prev.begin(), N, consume_result::accept_batch);
+    }
+
+    consume_result
+    accept_batch_start(const model::record_batch_header& h) const override {
+        auto ix = 0;
+        for (const auto& c : _cons) {
+            if (_prev[ix] != consume_result::stop_parser) {
+                auto r = c->accept_batch_start(h);
+                _prev[ix] = r;
+            }
+            ix++;
+        }
+        size_t n_stop = 0;
+        size_t n_skip = 0;
+        for (auto cr : _prev) {
+            switch (cr) {
+            case consume_result::accept_batch:
+                break;
+            case consume_result::skip_batch:
+                n_skip++;
+                break;
+            case consume_result::stop_parser:
+                n_stop++;
+                break;
+            }
+        }
+        if (n_stop == _prev.size()) {
+            return consume_result::stop_parser;
+        }
+        if (n_skip == _prev.size()) {
+            return consume_result::skip_batch;
+        }
+        return consume_result::accept_batch;
+    }
+
+    void consume_batch_start(
+      model::record_batch_header hdr,
+      size_t offset, // NOLINT
+      size_t size) override {
+        auto ix = 0;
+        for (const auto& c : _cons) {
+            if (_prev[ix] == consume_result::accept_batch) {
+                c->consume_batch_start(hdr, offset, size);
+            } else if (_prev[ix] == consume_result::skip_batch) {
+                c->skip_batch_start(hdr, offset, size);
+            }
+            ix++;
+        }
+    }
+
+    void skip_batch_start(
+      model::record_batch_header hdr,
+      size_t offset, // NOLINT
+      size_t size) override {
+        auto ix = 0;
+        for (const auto& c : _cons) {
+            if (_prev[ix] == consume_result::skip_batch) {
+                c->skip_batch_start(hdr, offset, size);
+            }
+            ix++;
+        }
+    }
+
+    void consume_records(iobuf&& b) override {
+        auto ix = 0;
+        for (auto& c : _cons) {
+            if (_prev[ix] == consume_result::accept_batch) {
+                c->consume_records(b.share(0, b.size_bytes()));
```

Comment:
why are you taking the iobuf by r-value reference, and then not actually taking ownership of it?